### PR TITLE
Support sha224, sha384, sha512 for hash() and hash_hmac().

### PR DESCRIPTION
**Pull request by AlexK0**
This patch enables `sha224`, `sha384`, `sha512` algorithms for `hash()` and `hash_hmac()`.
Also `hash_hmac_algos()` was implemented.

### CODE REVIEW

**Review by unserialize — APPROVED** [review]

File: runtime/openssl.cpp [hunk 1]
```diff
@@ -36,127 +37,110 @@
...
-      res.assign(64, false);
-    }
+struct HashTraits {
+private:
+  template<class F>
+  string call_hash_algo(bool raw_output, const F &hash_alho) const noexcept {
+    string res{hash_len * (raw_output ? 1 : 2), false};
 
-    dl::critical_section_call(SHA256, reinterpret_cast<const unsigned char *>(s.c_str()), s.size(), reinterpret_cast<unsigned char *>(res.buffer()));
+    hash_alho(res);
```
Comment (eugene536) [inline runtime/openssl.cpp:48]: nit) hash_АЛFА

File: runtime/openssl.cpp [hunk 2]
```diff
@@ -253,6 +237,18 @@ int64_t f$crc32_file(const string &file_name) {
   return res ^ std::numeric_limits<uint32_t>::max();
 }
 
+bool f$hash_equals(const string &known_string, const string &user_string) noexcept {
+  if (known_string.size() != user_string.size()) {
+    return false;
+  }
+  int result = 0;
+  // This is security sensitive code. Do not optimize this for speed
```
Comment (eugene536) [inline runtime/openssl.cpp:245]: nit) я бы еще вот этот коментал скопипастил
/* {{{ Compares two strings using the same time whether they're equal or not.
   A difference in length will leak */

Comment (unserialize) [reply runtime/openssl.cpp:245]: Sorry, we don't understand Russian in Team KPHP 😀

Comment (eugene536) [reply runtime/openssl.cpp:245]: LOL, I forgot about it

**Review by eugene536 — APPROVED** [review]

**Pull request merged**